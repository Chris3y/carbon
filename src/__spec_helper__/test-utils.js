const isUpper = char => char.toUpperCase() === char;
const humpToDash = (acc, char) => `${acc}${isUpper(char) ? `-${char.toLowerCase()}` : char}`;

export const toCSSCase = (str) => {
  return str.split('').reduce(humpToDash, '');
};

export const assertStyleMatch = (styleSpec, component) => {
  Object.entries(styleSpec).forEach(([attr, value]) => {
    expect(component).toHaveStyleRule(toCSSCase(attr), value);
  });
};
