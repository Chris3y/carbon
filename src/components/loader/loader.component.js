import React from 'react';
import PropTypes from 'prop-types';
import tagComponent from '../../utils/helpers/tags/tags';
import OptionsHelper from '../../utils/helpers/options-helper';
import StyledLoader from './loader.style';
import StyledLoaderSquare from './loader-square.style';

const Loader = (props) => {
  return (
    <StyledLoader
      size={ props.size }
      className={ props.className }
      isInsideButton={ props.isInsideButton }
      { ...tagComponent('loader', props) }
    >
      <StyledLoaderSquare isInsideButton={ props.isInsideButton } size={ props.size } />
      <StyledLoaderSquare isInsideButton={ props.isInsideButton } size={ props.size } />
      <StyledLoaderSquare isInsideButton={ props.isInsideButton } size={ props.size } />
    </StyledLoader>
  );
};

Loader.defaultProps = {
  size: 'small',
  isInsideButton: false
};

Loader.propTypes = {
  className: PropTypes.string,
  size: PropTypes.oneOf(OptionsHelper.sizesBinary),
  isInsideButton: PropTypes.bool
};

export default Loader;
