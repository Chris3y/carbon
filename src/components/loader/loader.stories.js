import React from 'react';
import { storiesOf } from '@storybook/react';
import { select, boolean } from '@storybook/addon-knobs';
import OptionsHelper from '../../utils/helpers/options-helper';
import Loader from './loader.component';
import Spinner from '../spinner/spinner.component';
import Button from '../button/button.component';
import { notes, info } from './documentation';

storiesOf('Loader', module)
  .add(
    'default',
    () => {
      const size = select('size', OptionsHelper.sizesBinary, Loader.defaultProps.size);
      const isInsideButton = boolean('isInsideButton', Loader.defaultProps.isInsideButton);

      return <Loader size={ size } isInsideButton={ isInsideButton } />;
    },
    {
      info: { text: info },
      notes: { markdown: notes }
    }
  )
  .add(
    'inside a button',
    () => {
      const size = select('size', OptionsHelper.sizesBinary, Loader.defaultProps.size);
      const isInsideButton = boolean('isInsideButton', true);

      return (
        <Button buttonType='primary'>
          <Loader size={ size } isInsideButton={ isInsideButton } />
        </Button>
      );
    },
    {
      info: { text: info },
      notes: { markdown: notes }
    }
  )
  .add(
    'legacy spinner',
    () => {
      const type = select('as', OptionsHelper.colors, Spinner.defaultProps.as);
      const size = select('size', OptionsHelper.sizesFull, Spinner.defaultProps.size);

      return <Spinner as={ type } size={ size } />;
    },
    {
      info: { text: info },
      notes: { markdown: notes }
    }
  );
