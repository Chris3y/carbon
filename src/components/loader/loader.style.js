import styled from 'styled-components';
import PropTypes from 'prop-types';
import baseTheme from '../../style/themes/base';

const StyledLoader = styled.div`
  padding: 0 3px 3px 0;
  text-align: center;
`;

StyledLoader.defaultProps = {
  theme: baseTheme
};

StyledLoader.propTypes = {
  theme: PropTypes.object
};

export default StyledLoader;
