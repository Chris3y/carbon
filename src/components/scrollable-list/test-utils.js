import React from 'react';
import ScrollableListItem from './scrollable-list-item.component';

const makeArray = n => [...Array(n).keys()];

const listItemReducer = ({ nonSelectables = [], customSelectables = [] }) => {
  // generate jsx for selectable list items, based on indexes in config
  return (acc, item, index) => {
    if (nonSelectables.includes(item)) return [...acc, <div />];
    if (customSelectables.includes(item)) return [...acc, <div isSelectable />];

    return [...acc, <ScrollableListItem id={ index } />];
  };
};

const renderListItems = (opts) => {
  return makeArray(opts.num).reduce(listItemReducer(opts), []);
};


const keyPress = (code) => {
  const ev = new KeyboardEvent('keydown', { which: code });
  document.dispatchEvent(ev);
};

const keyMap = {
  UpArrow: '38',
  DownArrow: '40',
  RightArrow: '39',
  LeftArrow: '37',
  Enter: '13',
  Tab: '9'
};

const repeat = action => (n = 1) => makeArray(n).forEach(() => action());

const keyboard = Object.keys(keyMap).reduce((acc, key) => {
  acc[`press${key}`] = () => repeat(keyPress(keyMap[key]));
  return acc;
}, {});

const listFrom = wrapper => wrapper.find('ul');
const childrenFrom = node => node.children();

const hoverList = wrapper => (item) => {
  childrenFrom(listFrom(wrapper)).at(item).simulate('mouseover');
};

const simulateEvent = eventName => wrapper => wrapper.simulate(eventName);
const click = simulateEvent('click');

const selectedItemOf = wrapper => wrapper.state().selectedItem;
const isUnique = (val, index, self) => self.indexOf(val) === index;
const isSelectableGiven = nonSelectables => i => !nonSelectables.includes(i);

const selectedItemReducer = method => wrapper => (acc, i) => {
  method(wrapper)(i);
  return [...acc, selectedItemOf(wrapper)];
};

const arraysEqual = (arr1, arr2) => arr1.sort().join(',') === arr2.sort().join(',');

const assertCorrectTraversal = method => expect => ({ num, nonSelectables = [] }) => (wrapper) => {
  const array = makeArray(num);
  const validIndexes = array.filter(isSelectableGiven(nonSelectables));

  const indexesThatWereSelected = array
    .reduce(selectedItemReducer(method)(wrapper), [])
    .filter(isUnique);
  expect(arraysEqual(validIndexes, indexesThatWereSelected)).toBeTruthy();
};

const assertKeyboardTraversal = assertCorrectTraversal(() => keyboard.pressDownArrow)(expect);
const assertHoverTraversal = assertCorrectTraversal(wrapper => hoverList(wrapper))(expect);

export {
  hoverList,
  selectedItemOf,
  childrenFrom,
  renderListItems,
  keyboard,
  assertKeyboardTraversal,
  assertHoverTraversal,
  listFrom,
  click
};
