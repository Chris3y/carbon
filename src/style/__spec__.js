import buildPalette from './palette';
import palette from './palette';
import mix from './utils/mix';
import config from './color-config';



const assertCorrectColorMix = (config, paletteObject) => {
  Object.keys(config).forEach(col => {
    const match = /([A-Za-z]+)([\d]?[\d]?)/.exec(col);

    const func = match[1], weight = Number(match[2]);

    if (!paletteObject[func]) return 

    expect(paletteObject[func](weight)).toEqual(config[col]);
  })
}


describe('functional palette', () => {
  let colorConfig = {
    brilliantGreenBase: '00DC00',
    brilliantGreenShade20: '00B000',
    brilliantGreenTint50: '7FED7F',
    goldBase: 'FFB500',
    goldTint50: 'FFDA7F',
    errorRedBase: 'C7384F',
    errorRedShade20: '9F2C3F',
    genericGreenTint50: '7FCC7F',
    genericGreenTint30: '4CB74C',
    genericGreenShade15: '008200',
    genericGreenShade35: '006300',
    genericGreenShade55: '004400',
    productGreenTint50: '7FD1BA',
    productGreenTint30: '4CBE9F',
    productGreenShade21: '00805D',
    productGreenShade41: '006045',
    productGreenShade61: '003F2E',
    productBlueTint50: '7FBBE3',
    productBlueTint30: '4C9FD8',
    productBlueShade3: '0073C2',
    productBlueShade23: '005B9A',
    productBlueShade43: '004372',
    amethystTint50: 'AB95C1',
    amethystTint30: '8A6BA8',
    amethystTint10: '68418F',
    amethystBase: '582C83',
    amethystShade10: '4F2775',
    amethystShade30: '3D1E5B',
    slateTint95: 'F2F4F5',
    slateTint90: 'E5EAEC',
    slateTint80: 'CCD6DA',
    slateTint60: '99ADB6',
    slateTint40: '668491',
    slateTint20: '335B6D',
    slateShade60: '00141D'
  }

  it('produces the correct color mix', () => {
    assertCorrectColorMix(colorConfig, palette(config))
  })
})
