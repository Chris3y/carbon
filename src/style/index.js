import generatePalette from './palette';
import atOpacity from './utils/at_opacity';

const blackAtOpacity = atOpacity('#FFFFFF');
const addOpacity = (color, opacity) => atOpacity(color)(opacity);

export {
  addOpacity,
  blackAtOpacity,
  generatePalette
};
