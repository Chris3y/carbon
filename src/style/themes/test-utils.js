import generatePalette from '../palette';
import atOpacity from '../utils/at-opacity';
import baseColors from '../color-config';

const addHex = base => `#${base}`;

export const palette = {
  ...generatePalette(baseColors),
  atOpacity,
  ...Object.keys(baseColors).reduce((acc, col) => {
    acc[col] = addHex(baseColors[col]);
    return acc;
  }, {})
};
const blackWithOpacity = atOpacity('#000000');

export const baseThemeConfig = {
  colors: {
    brand: palette.brilliantGreen,
    white: '#FFFFFF',
    error: palette.errorRed,
    warning: palette.gold,
    success: palette.brilliantGreenShade(20),
    info: palette.productBlueShade(3),
    text: {
      body: blackWithOpacity(0.9),
      disabled: blackWithOpacity(0.55),
      placeholder: blackWithOpacity(0.3)
    }
  }
};


const isObject = (obj) => {
  return typeof obj === 'object' && obj !== null;
};

export const assertIsSubset = (obj, comparison) => {
  if (!isObject(obj)) {
    // no further nesting, assert that values are equal
    expect(obj).toEqual(comparison);
    return;
  }

  const objKeys = Object.keys(obj);
  const comparisonKeys = Object.keys(comparison);

  objKeys.forEach((key) => {
    // assert that keys are present
    expect(comparisonKeys.includes(key)).toBeTruthy();

    // repeat for nested objects
    assertIsSubset(obj[key], comparison[key]);
  });
};
