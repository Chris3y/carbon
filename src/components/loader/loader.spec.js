import React from 'react';
import TestRenderer from 'react-test-renderer';
import 'jest-styled-components';
import { shallow } from 'enzyme';
import { ThemeProvider } from 'emotion-theming';
import Loader from './loader.component';
import { StyledLoader } from './loader.style';
import OptionsHelper from '../../utils/helpers/options-helper';
import { assertStyleMatch } from '../../__spec_helper__/test-utils';
import baseTheme from '../../style/themes/base';

function render(props) {
  return shallow(<Loader { ...props } />);
}

function renderStyles(props) {
  return TestRenderer.create(<StyledLoader { ...props } />);
}

describe('Loader', () => {
  let wrapper;
  it('renders as expected', () => {
    wrapper = render();
    expect(wrapper).toMatchSnapshot();
  });

  it('contains custom className if passed as a prop', () => {
    wrapper = render({ className: 'class' });
    expect(wrapper.find('.class').exists()).toEqual(true);
  });

  it('applies proper color', () => {
    wrapper = renderStyles({ isInsideButton: true });
    assertStyleMatch(
      {
        backgroundColor: baseTheme.colors.white
      },
      wrapper.toJSON()
    );
  });

  it('applies width, height and border-width dependent on the passed size', () => {});
});
