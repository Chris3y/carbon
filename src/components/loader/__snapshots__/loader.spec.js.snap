// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader renders as expected 1`] = `
<styled.div
  data-component="loader"
  isInsideButton={false}
  size="small"
  theme={
    Object {
      "colors": Object {
        "base": "#00A376",
        "border": "#668491",
        "brand": "#00DC00",
        "destructive": Object {
          "hover": "#9F2C3F",
        },
        "error": "#C7384F",
        "focus": "#FFB500",
        "focusedIcon": "#335B6D",
        "focusedLinkBackground": "#FFDA7F",
        "hoveredTabKeyline": "#4CB74C",
        "info": "#0073C2",
        "primary": "#26A826",
        "secondary": "#006300",
        "success": "#00B000",
        "tertiary": "#004400",
        "warning": "#E96400",
        "white": "#FFFFFF",
        "whiteMix": "#E5F4E5",
      },
      "disabled": Object {
        "background": "#E5EAEC",
        "border": "#CCD6DA",
        "button": "#E5EAEC",
        "buttonText": "rgba(0,0,0,.2)",
        "disabled": "rgba(0,0,0,0.55)",
        "input": "#F2F4F5",
        "text": "rgba(0,0,0,0.3)",
      },
      "icon": Object {
        "focus": "#335B6D",
      },
      "name": "base",
      "shadows": Object {
        "cards": "0 3px 3px 0 rgba(0,20,29,0.2), 0 2px 4px 0 rgba(0,20,29,0.15)",
        "depth1": "0 5px 5px 0 rgba(0,20,29,0.2), 0 10px 10px 0 rgba(0,20,29,0.1)",
        "depth2": "0 10px 20px 0 rgba(0,20,29,0.2), 0 20px 40px 0 rgba(0,20,29,0.1)",
        "depth3": "0 10px 30px 0 rgba(0,20,29,0.1), 0 30px 60px 0 rgba(0,20,29,0.1)",
        "depth4": "0 10px 40px 0 rgba(0,20,29,0.04), 0 50px 80px 0 rgba(0,20,29,0.1)",
      },
      "text": Object {
        "color": "rgba(0,0,0,0.9)",
        "placeholder": "rgba(0,0,0,0.3)",
        "size": "14px",
      },
    }
  }
>
  <styled.div
    isInsideButton={false}
    size="small"
    theme={
      Object {
        "colors": Object {
          "base": "#00A376",
          "border": "#668491",
          "brand": "#00DC00",
          "destructive": Object {
            "hover": "#9F2C3F",
          },
          "error": "#C7384F",
          "focus": "#FFB500",
          "focusedIcon": "#335B6D",
          "focusedLinkBackground": "#FFDA7F",
          "hoveredTabKeyline": "#4CB74C",
          "info": "#0073C2",
          "primary": "#26A826",
          "secondary": "#006300",
          "success": "#00B000",
          "tertiary": "#004400",
          "warning": "#E96400",
          "white": "#FFFFFF",
          "whiteMix": "#E5F4E5",
        },
        "disabled": Object {
          "background": "#E5EAEC",
          "border": "#CCD6DA",
          "button": "#E5EAEC",
          "buttonText": "rgba(0,0,0,.2)",
          "disabled": "rgba(0,0,0,0.55)",
          "input": "#F2F4F5",
          "text": "rgba(0,0,0,0.3)",
        },
        "icon": Object {
          "focus": "#335B6D",
        },
        "name": "base",
        "shadows": Object {
          "cards": "0 3px 3px 0 rgba(0,20,29,0.2), 0 2px 4px 0 rgba(0,20,29,0.15)",
          "depth1": "0 5px 5px 0 rgba(0,20,29,0.2), 0 10px 10px 0 rgba(0,20,29,0.1)",
          "depth2": "0 10px 20px 0 rgba(0,20,29,0.2), 0 20px 40px 0 rgba(0,20,29,0.1)",
          "depth3": "0 10px 30px 0 rgba(0,20,29,0.1), 0 30px 60px 0 rgba(0,20,29,0.1)",
          "depth4": "0 10px 40px 0 rgba(0,20,29,0.04), 0 50px 80px 0 rgba(0,20,29,0.1)",
        },
        "text": Object {
          "color": "rgba(0,0,0,0.9)",
          "placeholder": "rgba(0,0,0,0.3)",
          "size": "14px",
        },
      }
    }
  />
  <styled.div
    isInsideButton={false}
    size="small"
    theme={
      Object {
        "colors": Object {
          "base": "#00A376",
          "border": "#668491",
          "brand": "#00DC00",
          "destructive": Object {
            "hover": "#9F2C3F",
          },
          "error": "#C7384F",
          "focus": "#FFB500",
          "focusedIcon": "#335B6D",
          "focusedLinkBackground": "#FFDA7F",
          "hoveredTabKeyline": "#4CB74C",
          "info": "#0073C2",
          "primary": "#26A826",
          "secondary": "#006300",
          "success": "#00B000",
          "tertiary": "#004400",
          "warning": "#E96400",
          "white": "#FFFFFF",
          "whiteMix": "#E5F4E5",
        },
        "disabled": Object {
          "background": "#E5EAEC",
          "border": "#CCD6DA",
          "button": "#E5EAEC",
          "buttonText": "rgba(0,0,0,.2)",
          "disabled": "rgba(0,0,0,0.55)",
          "input": "#F2F4F5",
          "text": "rgba(0,0,0,0.3)",
        },
        "icon": Object {
          "focus": "#335B6D",
        },
        "name": "base",
        "shadows": Object {
          "cards": "0 3px 3px 0 rgba(0,20,29,0.2), 0 2px 4px 0 rgba(0,20,29,0.15)",
          "depth1": "0 5px 5px 0 rgba(0,20,29,0.2), 0 10px 10px 0 rgba(0,20,29,0.1)",
          "depth2": "0 10px 20px 0 rgba(0,20,29,0.2), 0 20px 40px 0 rgba(0,20,29,0.1)",
          "depth3": "0 10px 30px 0 rgba(0,20,29,0.1), 0 30px 60px 0 rgba(0,20,29,0.1)",
          "depth4": "0 10px 40px 0 rgba(0,20,29,0.04), 0 50px 80px 0 rgba(0,20,29,0.1)",
        },
        "text": Object {
          "color": "rgba(0,0,0,0.9)",
          "placeholder": "rgba(0,0,0,0.3)",
          "size": "14px",
        },
      }
    }
  />
  <styled.div
    isInsideButton={false}
    size="small"
    theme={
      Object {
        "colors": Object {
          "base": "#00A376",
          "border": "#668491",
          "brand": "#00DC00",
          "destructive": Object {
            "hover": "#9F2C3F",
          },
          "error": "#C7384F",
          "focus": "#FFB500",
          "focusedIcon": "#335B6D",
          "focusedLinkBackground": "#FFDA7F",
          "hoveredTabKeyline": "#4CB74C",
          "info": "#0073C2",
          "primary": "#26A826",
          "secondary": "#006300",
          "success": "#00B000",
          "tertiary": "#004400",
          "warning": "#E96400",
          "white": "#FFFFFF",
          "whiteMix": "#E5F4E5",
        },
        "disabled": Object {
          "background": "#E5EAEC",
          "border": "#CCD6DA",
          "button": "#E5EAEC",
          "buttonText": "rgba(0,0,0,.2)",
          "disabled": "rgba(0,0,0,0.55)",
          "input": "#F2F4F5",
          "text": "rgba(0,0,0,0.3)",
        },
        "icon": Object {
          "focus": "#335B6D",
        },
        "name": "base",
        "shadows": Object {
          "cards": "0 3px 3px 0 rgba(0,20,29,0.2), 0 2px 4px 0 rgba(0,20,29,0.15)",
          "depth1": "0 5px 5px 0 rgba(0,20,29,0.2), 0 10px 10px 0 rgba(0,20,29,0.1)",
          "depth2": "0 10px 20px 0 rgba(0,20,29,0.2), 0 20px 40px 0 rgba(0,20,29,0.1)",
          "depth3": "0 10px 30px 0 rgba(0,20,29,0.1), 0 30px 60px 0 rgba(0,20,29,0.1)",
          "depth4": "0 10px 40px 0 rgba(0,20,29,0.04), 0 50px 80px 0 rgba(0,20,29,0.1)",
        },
        "text": Object {
          "color": "rgba(0,0,0,0.9)",
          "placeholder": "rgba(0,0,0,0.3)",
          "size": "14px",
        },
      }
    }
  />
</styled.div>
`;
